src/update-done/update-done.c

```diff
@@ -76,7 +77,13 @@ static int apply_timestamp(const char *path, struct timespec *ts) {
                 return log_error_errno(r, "Failed to write timestamp file: %m");
 
         if (futimens(fd, twice) < 0)
-                return log_error_errno(errno, "Failed to update timestamp on %s: %m", path);
+                return log_error_errno(errno, "Failed to update timestamp on %s: %m", tmp);
+
+        /* fix permissions */
+        (void) fchmod(fd, 0644);
+        r = link_tmpfile(fd, tmp, path);
+        if (r < 0)
+                return log_error_errno(r, "Failed to moved update timestamp from  %s: %m", tmp);
```

Comment:
If you use the cleanup stuff suggested above, you'd need to explicitly free "tmp" after this, and set it to NULL, so that the cleanup logic won't try to unlink() it here. Specifically this should enough to add here:

```c
tmp = mfree(tmp);
```